org.objectweb.asm.ClassTooLargeException when building site

I'm trying to migrate a site to Roq, and I'm hitting a very strange error. Before I share that, you can find the current (as of this writing) state of the migration here: https://github.com/jasondlee/wildfly-elytron/tree/CLASS_TOO_LARGE/

And the stacktrace:

```
2025-04-23 15:04:22,682 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.MainClassBuildStep#build threw an exception: org.objectweb.asm.ClassTooLargeException: Class too large: io/quarkus/runner/recorded/RoqGeneratorProcessor$initHandler1143282271
	at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:616)
	at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:246)
	at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:257)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:583)
	at io.quarkus.deployment.steps.MainClassBuildStep.writeRecordedBytecode(MainClassBuildStep.java:512)
	at io.quarkus.deployment.steps.MainClassBuildStep.build(MainClassBuildStep.java:211)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:499)

	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:372)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:289)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.createInitialRuntimeApplication(AugmentActionImpl.java:61)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.firstStart(IsolatedDevModeMain.java:89)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:428)
	at io.quarkus.deployment.dev.IsolatedDevModeMain.accept(IsolatedDevModeMain.java:55)
	at io.quarkus.bootstrap.app.CuratedApplication.runInCl(CuratedApplication.java:138)
	at io.quarkus.bootstrap.app.CuratedApplication.runInAugmentClassLoader(CuratedApplication.java:93)
	at io.quarkus.deployment.dev.DevModeMain.start(DevModeMain.java:107)
	at io.quarkus.deployment.dev.DevModeMain.main(DevModeMain.java:70)
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.MainClassBuildStep#build threw an exception: org.objectweb.asm.ClassTooLargeException: Class too large: io/quarkus/runner/recorded/RoqGeneratorProcessor$initHandler1143282271
	at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:616)
	at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:246)
	at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:257)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:583)
	at io.quarkus.deployment.steps.MainClassBuildStep.writeRecordedBytecode(MainClassBuildStep.java:512)
	at io.quarkus.deployment.steps.MainClassBuildStep.build(MainClassBuildStep.java:211)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:499)

	at io.quarkus.builder.Execution.run(Execution.java:122)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:78)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:161)
	at io.quarkus.runner.bootstrap.AugmentActionImpl.runAugment(AugmentActionImpl.java:368)
	... 9 more
Caused by: org.objectweb.asm.ClassTooLargeException: Class too large: io/quarkus/runner/recorded/RoqGeneratorProcessor$initHandler1143282271
	at org.objectweb.asm.ClassWriter.toByteArray(ClassWriter.java:616)
	at io.quarkus.gizmo.ClassCreator.writeTo(ClassCreator.java:246)
	at io.quarkus.gizmo.ClassCreator.close(ClassCreator.java:257)
	at io.quarkus.deployment.recording.BytecodeRecorderImpl.writeBytecode(BytecodeRecorderImpl.java:583)
	at io.quarkus.deployment.steps.MainClassBuildStep.writeRecordedBytecode(MainClassBuildStep.java:512)
	at io.quarkus.deployment.steps.MainClassBuildStep.build(MainClassBuildStep.java:211)
	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2675)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2654)
	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1627)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1594)
	at java.base/java.lang.Thread.run(Thread.java:1583)
	at org.jboss.threads.JBossThread.run(JBossThread.java:499)
```

To reproduce, just run `quarkus dev` with that tag checked out.

--- a/docs/modules/ROOT/pages/_includes/quarkus-roq-frontmatter.adoc
+++ b/docs/modules/ROOT/pages/_includes/quarkus-roq-frontmatter.adoc
@@ -408,9 +408,9 @@ endif::add-copy-button-to-env-var[]
 |string
 |
 
-a|icon:lock[title=Fixed at build time] [[quarkus-roq-frontmatter_site-base-path]] [.property-path]##link:#quarkus-roq-frontmatter_site-base-path[`site.base-path`]##
+a|icon:lock[title=Fixed at build time] [[quarkus-roq-frontmatter_site-path-prefix]] [.property-path]##link:#quarkus-roq-frontmatter_site-path-prefix[`site.path-prefix`]##
 ifdef::add-copy-button-to-config-props[]
-config_property_copy_button:+++site.base-path+++[]
+config_property_copy_button:+++site.path-prefix+++[]
 endif::add-copy-button-to-config-props[]
 
 
@@ -418,14 +418,14 @@ endif::add-copy-button-to-config-props[]
 --
 *READ CAREFULLY:* +
 The root path of your site (e.g. `/blog`) should be set using `quarkus.http.root-path`. +
-This path should be relative to the Quarkus HTTP root path and is meant to be used only when the Roq site is served alongside a Quarkus application.
+This path prefix should be relative to the Quarkus HTTP root path and is meant to be used only when the Roq site is served alongside a Quarkus application on a separate path.
 
 
 ifdef::add-copy-button-to-env-var[]
-Environment variable: env_var_with_copy_button:+++SITE_BASE_PATH+++[]
+Environment variable: env_var_with_copy_button:+++SITE_PATH_PREFIX+++[]
 endif::add-copy-button-to-env-var[]
 ifndef::add-copy-button-to-env-var[]
-Environment variable: `+++SITE_BASE_PATH+++`
+Environment variable: `+++SITE_PATH_PREFIX+++`
 endif::add-copy-button-to-env-var[]
 --
 |string

--- a/docs/modules/ROOT/pages/_includes/quarkus-roq-frontmatter_site.adoc
+++ b/docs/modules/ROOT/pages/_includes/quarkus-roq-frontmatter_site.adoc
@@ -408,9 +408,9 @@ endif::add-copy-button-to-env-var[]
 |string
 |
 
-a|icon:lock[title=Fixed at build time] [[quarkus-roq-frontmatter_site-base-path]] [.property-path]##link:#quarkus-roq-frontmatter_site-base-path[`site.base-path`]##
+a|icon:lock[title=Fixed at build time] [[quarkus-roq-frontmatter_site-path-prefix]] [.property-path]##link:#quarkus-roq-frontmatter_site-path-prefix[`site.path-prefix`]##
 ifdef::add-copy-button-to-config-props[]
-config_property_copy_button:+++site.base-path+++[]
+config_property_copy_button:+++site.path-prefix+++[]
 endif::add-copy-button-to-config-props[]
 
 
@@ -418,14 +418,14 @@ endif::add-copy-button-to-config-props[]
 --
 *READ CAREFULLY:* +
 The root path of your site (e.g. `/blog`) should be set using `quarkus.http.root-path`. +
-This path should be relative to the Quarkus HTTP root path and is meant to be used only when the Roq site is served alongside a Quarkus application.
+This path prefix should be relative to the Quarkus HTTP root path and is meant to be used only when the Roq site is served alongside a Quarkus application on a separate path.
 
 
 ifdef::add-copy-button-to-env-var[]
-Environment variable: env_var_with_copy_button:+++SITE_BASE_PATH+++[]
+Environment variable: env_var_with_copy_button:+++SITE_PATH_PREFIX+++[]
 endif::add-copy-button-to-env-var[]
 ifndef::add-copy-button-to-env-var[]
-Environment variable: `+++SITE_BASE_PATH+++`
+Environment variable: `+++SITE_PATH_PREFIX+++`
 endif::add-copy-button-to-env-var[]
 --
 |string

--- a/roq-generator/deployment/src/main/java/io/quarkiverse/roq/generator/deployment/BuildSelectionBuildItem.java
+++ b/roq-generator/deployment/src/main/java/io/quarkiverse/roq/generator/deployment/BuildSelectionBuildItem.java
@@ -1,0 +1,25 @@
+package io.quarkiverse.roq.generator.deployment;
+
+import java.util.Map;
+
+import io.quarkiverse.roq.generator.runtime.RoqSelection;
+import io.quarkiverse.roq.generator.runtime.StaticFile;
+import io.quarkus.builder.item.SimpleBuildItem;
+
+public final class BuildSelectionBuildItem extends SimpleBuildItem {
+    private final Map<String, StaticFile> staticFiles;
+    private final RoqSelection selectedPaths;
+
+    public BuildSelectionBuildItem(Map<String, StaticFile> staticFiles, RoqSelection selectedPaths) {
+        this.staticFiles = staticFiles;
+        this.selectedPaths = selectedPaths;
+    }
+
+    public Map<String, StaticFile> staticFiles() {
+        return staticFiles;
+    }
+
+    public RoqSelection selectedPaths() {
+        return selectedPaths;
+    }
+}

--- a/roq-generator/deployment/src/main/java/io/quarkiverse/roq/generator/deployment/RoqGeneratorProcessor.java
+++ b/roq-generator/deployment/src/main/java/io/quarkiverse/roq/generator/deployment/RoqGeneratorProcessor.java
@@ -65,15 +65,12 @@ class RoqGeneratorProcessor {
     }
 
     @BuildStep
-    @Record(ExecutionTime.STATIC_INIT)
-    void initHandler(
+    public BuildSelectionBuildItem initBuildSelection(
             RoqGeneratorConfig config,
             List<GeneratedStaticResourceBuildItem> generatedStaticResources,
             List<NotFoundPageDisplayableEndpointBuildItem> notFoundPageDisplayableEndpoints,
             List<SelectedPathBuildItem> selectedPaths,
-            StaticResourcesBuildItem staticResourcesBuildItem,
-            OutputTargetBuildItem outputTarget,
-            RoqGeneratorRecorder recorder) {
+            StaticResourcesBuildItem staticResourcesBuildItem) {
         Set<String> staticPaths = new HashSet<>();
         if (staticResourcesBuildItem != null) {
             staticPaths.addAll(staticResourcesBuildItem.getPaths().stream().map(PathUtils::prefixWithSlash).toList());
@@ -92,8 +89,24 @@ class RoqGeneratorProcessor {
         final Map<String, StaticFile> staticFiles = new HashMap<>();
         final RoqSelection buildSelectedPaths = getSelectedPaths(config, selectedPathsFromBuildItems,
                 generatedStaticResourcesMap, staticPaths, staticFiles);
-        recorder.setBuildSelectedPaths(buildSelectedPaths);
-        recorder.setStaticFiles(staticFiles);
+        return new BuildSelectionBuildItem(staticFiles, buildSelectedPaths);
+    }
+
+    @BuildStep
+    @Record(ExecutionTime.STATIC_INIT)
+    void initStaticFiles(
+            BuildSelectionBuildItem buildSelection,
+            RoqGeneratorRecorder recorder) {
+        recorder.setStaticFiles(buildSelection.staticFiles());
+    }
+
+    @BuildStep
+    @Record(ExecutionTime.STATIC_INIT)
+    void initSelection(
+            BuildSelectionBuildItem buildSelection,
+            OutputTargetBuildItem outputTarget,
+            RoqGeneratorRecorder recorder) {
+        recorder.setBuildSelectedPaths(buildSelection.selectedPaths());
         recorder.setOutputTarget(outputTarget.getOutputDirectory().toAbsolutePath().toString());
     }
 

